x/fswap/keeper/keeper.go

```diff
@@ -82,6 +105,63 @@ func (k Keeper) Swap(ctx sdk.Context, addr sdk.AccAddress, fromCoinAmount sdk.Co
 	return nil
 }
 
+func (k Keeper) SetSwap(ctx sdk.Context, swap types.Swap, toDenomMetadata bank.Metadata) error {
+	isNewSwap := true
+	if _, err := k.getSwap(ctx, swap.FromDenom, swap.ToDenom); err == nil {
+		isNewSwap = false
+	}
+
+	if !isNewSwap && !k.config.UpdateAllowed {
+		return sdkerrors.ErrInvalidRequest.Wrap("update existing swap not allowed")
+	}
+
+	if isNewSwap {
+		if err := k.increaseSwapCount(ctx); err != nil {
+			return err
+		}
+	}
+
+	stats, err := k.getSwapStats(ctx)
+	if err != nil {
+		return err
+	}
+
+	if int(stats.SwapCount) > k.config.MaxSwaps && !k.isUnlimited() {
+		return types.ErrCanNotHaveMoreSwap.Wrapf("cannot make more swaps, max swaps is %d", k.config.MaxSwaps)
+	}
+
+	if isNewSwap {
+		swapped := types.Swapped{
+			FromCoinAmount: sdk.Coin{
+				Denom:  swap.GetFromDenom(),
+				Amount: sdk.ZeroInt(),
+			},
+			ToCoinAmount: sdk.Coin{
+				Denom:  swap.GetToDenom(),
+				Amount: sdk.ZeroInt(),
+			},
+		}
+		if err := k.setSwapped(ctx, swapped); err != nil {
+			return err
+		}
+	}
+
+	if err := k.setSwap(ctx, swap); err != nil {
+		return err
+	}
+
+	existingMetadata, ok := k.GetDenomMetaData(ctx, swap.ToDenom)
+	if !ok {
+		k.SetDenomMetaData(ctx, toDenomMetadata)
+		return nil
+	}
+	if !denomMetadataEqual(existingMetadata, toDenomMetadata) {
+		return sdkerrors.ErrInvalidRequest.Wrap("changing existing metadata not allowed")
+	}
```

Comment:
Is it better to move this part before creating swapped (before line133). The denomData can be created first.

Reply:
I'm sorry but I don't see any good point to move that part.